oxo.py

```diff
@@ -1,26 +1,90 @@
+
+        # Create the 2-dimensional array for the grid of squares
+        self.grid = [[0 for y in xrange(self.num_rows)] for x in xrange(self.num_cols)]
 
     def get_square(self, x, y):
         """ Return 0, 1 or 2 depending on the contents of the specified square. """
-        raise NotImplementedError("TODO: implement get_square")
+        return self.grid[x][y]
 
     def set_square(self, x, y, mark):
         """ If the specified square is currently empty (0), fill it with mark and return True.
-            If the square is not empty, leave it as-is and return False. """
-        raise NotImplementedError("TODO: implement set_square")
+            If the square is not empty, leave it as-is and return False.
+        """
+        if self.grid[x][y] is OxoTile.EMPTY:
+            self.grid[x][y] = mark
+            return True
+        else:
+            return False
 
     def is_board_full(self):
         """ If there are still empty squares on the board, return False.
-            If there are no empty squares, return True. """
-        raise NotImplementedError("TODO: implement is_board_full")
+            If there are no empty squares, return True.
+        """
+        # Return whether an empty square was (not!) found in any of the grid's column arrays
+        return not any(OxoTile.EMPTY in column for column in self.grid)
 
     def get_winner(self):
-        """ If a player has three in a row, return 1 or 2 depending on which player.
-            Otherwise, return 0. """
-        raise NotImplementedError("TODO: implement get_winner")
+        """ If a player has [winning_row_length] in a row, return 1 or 2 depending on which player.
+            Otherwise, return 0.
+        """
+        # Declare a search direction list, representing the XY offsets of
+        # up-right, right, down-right and down directions. These are the
+        # directions we'll search the grid along for matching O's or X's
+        dir_xy_offsets = [(1, 0), (1, 1), (1, -1), (0, 1)]
+
+        # Iterate through all squares in the grid and search each possible line direction
+        for start_x in xrange(0, self.num_cols):
+            for start_y in xrange(0, self.num_rows):
+                compare_type = self.grid[start_x][start_y]
+
+                # Don't search along empty tiles
+                if compare_type == OxoTile.EMPTY:
+                    continue
+
+                # Search along the four directions to see if there's a valid row
+                for offset in dir_xy_offsets:
+                    for step in xrange(1, self.winning_row_length):
+                        # Step
+                        current_x = start_x + offset[0] * step
+                        current_y = start_y + offset[1] * step
+
+                        # Don't go past the border!
+                        if current_x >= self.num_cols or current_y >= self.num_rows or \
+                                current_y < 0:
```

Comment:
It's not immediately obvious why you don't need to check `current_x < 0` (because `dir_xy_offsets` never specifies a negative x offset) -- worth adding a comment to this effect, or just add the check to remain consistent